lib/active_model/serializer/adapter/json_api/deserialization.rb

```diff
@@ -0,0 +1,120 @@
+module ActiveModel
+  class Serializer
+    module Adapter
+      class JsonApi
+        module Deserialization
+          module_function
+
+          # Parse a Hash or ActionController::Parameters representing a JSON API document
+          # into an ActiveRecord-ready hash.
+          #
+          # @param [Hash|Object implementing `to_h`] document
+          # @param [Hash] options
+          #   fields: Array of symbols and a Hash. Specify whitelisted fields, optionally
+          #     specifying the attribute name on the model.
+          # @return [Hash] ActiveRecord-ready hash
+          #
+          def parse(document, options = {})
+            fields = parse_fields(options[:fields])
+
+            hash = {}
+
+            document = document.to_h unless document.is_a?(Hash)
+            primary_data = document.fetch('data', {})
+            included_data = document.fetch('included', {})
+
+            hash.merge!(parse_primary_data(primary_data, fields))
+            hash.merge!(parse_included_data(included_data, fields))
+
+            hash
+          end
+
+          def parse_included_data(included_data, fields)
+            data = {}
+            included_data.each do |included_resource|
+              parsed = parse_primary_data(included_resource, fields)
+              # parsed should only have one key
+              # so there is no possibility of missing other keys when
+              # calling first
+              # NOTE: key is always going to be singular
+              key = parsed.keys.first
+              plural_key = key.pluralize
+              parsed_attributes = parsed[key]
+              # on multiple iterations of parsing the same type
+              # the key in data is changed to be plural,
+              # so we need to check for the existence of both
+              # singular and plural keys
+              exists = data.keys.include?(key) || data.keys.include?(plural_key)
```

Comment:
As someone who has looked a lot in the code of this project and monkey patched some of it to adapt in my needs I have to admit such comments would help me a lot + would make me confident to brag about my changes and send pull requests back. Now I feel like I have found a way to "make it work" how I want it but there might be a better way because I don't have a clue what the original maintainer meant/wanted to achieve on some parts. Ruby is awesome but sometimes it's not self explained like other languages :)